### PR TITLE
Use OpenAPI to document the API

### PR DESCRIPTION
**Pull request by jerivas**
## Description

Use OpenAPI to document the Metecho API. This replaces manually-maintained `.rst` files for each endpoint which were often outdated and inaccurate with an auto-generated `schema.yml` file:

- Extract API schema using [drf-spectacular](https://github.com/tfranzel/drf-spectacular)
- Generate browsable docs using [Redoc](https://redoc.ly/redoc/)

## Steps to test/reproduce

**Quick demo: use an [online Redoc viewer](http://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/oddbird/Metecho/automated-api-docs/docs/api/schema.yml) to parse the contents of [schema.yml](https://raw.githubusercontent.com/oddbird/Metecho/25bfc765803dc5ad998f0ff3738541017f9a58b2/docs/api/schema.yml).**

### Redoc viewer (same as above but running locally)
- Go to http://localhost:8000/api/schema/redoc/. Refresh the page whenever the views/serializers change.

### Updating the schema file

This is not required to generate the docs, but to ensure other tooling has an up to date version of the schema. Run `python manage.py spectacular --validate --file docs/api/schema.yml` to compile a new schema after changing views/serializers. This can be added to CI checks as well to verify the schema is always up to date.

## Show me
image

## What's the maintenance cost?

1. All CRUD actions on viewsets are documented automatically, they don't need any changes made to their view or serializer code.
2. Special actions on viewsets (any method decorated with `@action`) need to be decorated with `@extend_schema(request=InputSerializer, responses=OutputSerializer)`. This also means we should use more serializers instead of manually handling request data in the view (which I think is a good thing). The docstring will also be picked up by the schema. More advanced use cases are supported (like multiple return codes) but this should cover around 80% of cases.
3. Any custom serializer fields should be decorated with `@extend_schema_field(RETURN_TYPE)` if the return type cannot be derived from annotations. This also applies to serializer method fields.
4. If a serializer field simply calls a model method, the return type should be type-annotated in the model.

I think this PR already did the most difficult part, which is decorating and annotating all existing code. Maintenance going forward should be pretty easy and I would argue that it'll take less time than writing the docs manually, as well as being much less error-prone.

## More features

Since we now have a standard-compliant `schema.yml` file we can use it to auto-generate TypeScript interfaces and even an API client using something like [openapi-typescript-codegen](https://github.com/ferdikoomen/openapi-typescript-codegen). drf-spectacular mentions this use case and makes some recommendations in their [docs](https://drf-spectacular.readthedocs.io/en/latest/client_generation.html).

### CODE REVIEW

**Comment by jerivas** [comment]
[Automatically document the API](https://trello.com/c/nDfv90Wo/296-automatically-document-the-api)

**Review by jgerigmeyer — APPROVED** [review]
> This can be added to CI checks as well to verify the schema is always up to date.

I think we might as well go ahead and do this in this PR?

> As an alternative we could embed the Redoc viewer directly in the Sphinx-generated HTML...

I'm not clear on the pros/cons here, so I'll defer to whatever you recommend @jerivas 

> Since we now have a standard-compliant schema.yml file we can use it to auto-generate TypeScript interfaces and even an API client...

Added a story: https://trello.com/c/8HsEbszg

**Comment by jerivas** [comment]
> > This can be added to CI checks as well to verify the schema is always up to date.
> 
> I think we might as well go ahead and do this in this PR?

Will do 👍

> > As an alternative we could embed the Redoc viewer directly in the Sphinx-generated HTML...
> 
> I'm not clear on the pros/cons here, so I'll defer to whatever you recommend @jerivas

`sphinxcontrib-openapi` will just output some static HTML ([demo here](https://sphinxcontrib-openapi.readthedocs.io/#get--persons)), while [sphinxcontrib-redoc](https://sphinxcontrib-redoc.readthedocs.io/en/stable/) will generate and embed the Redoc UI in the HTML generated by Sphinx. My preference is towards Redoc. @davisagli does that work for you?

**Comment by jerivas** [comment]
@jgerigmeyer https://github.com/oddbird/Metecho/pull/1495/commits/9ecd8c77d00bda2d5f883bd4d69614dcf0b08a86 checks the schema as part of the python test suite.

**Review by jgerigmeyer — APPROVED** [review]

File: metecho/api/serializers.py [hunk 1]
```diff
@@ -103,26 +125,27 @@ class Meta:
...
-class ProjectSerializer(serializers.ModelSerializer):
-    id = serializers.CharField(read_only=True)
+class ProjectSerializer(HashIdModelSerializer):
+    slug = serializers.CharField()
+    old_slugs = StringListField(read_only=True)
     description_rendered = MarkdownField(source="description", read_only=True)
     repo_url = serializers.SerializerMethodField()
     repo_image_url = serializers.SerializerMethodField()
     has_push_permission = serializers.SerializerMethodField()
+    github_users = StringListField()
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:147]: @jerivas This introduced a bug I didn't notice earlier. `project.github_users` is a list of JSON objects, not a list of strings.

cc @dvdherron -- thanks for catching this.

Comment (jerivas) [reply metecho/api/serializers.py:147]: Thanks, working on a fix now

File: metecho/api/serializers.py [hunk 2]
```diff
@@ -103,26 +125,27 @@ class Meta:
...
+    org_config_names = StringListField(read_only=True)
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:148]: @jerivas Same bug here for `org_config_names`

**Pull request merged**